File: 1-two-sum/two-sum.py
# Optimal Code
class Solution:
    def twoSum(self, nums: list[int], target: int) -> list[int]:
        # dictionary to store : number --> index
        mp = {}
        # traverse the array
        for i in range(len(nums)):
            # calculate the number needed to reach the target
            complement = target - nums[i]
            # if the complement has alredy been seen return its index and the current index
            if complement in mp:
                return [mp[complement], i]
            # store the current number along with its index
            mp[nums[i]] = i
        # no valid pair found
        return []

# Time Complexity : O(N)
# Space Complexity : O(N)

File: 1-two-sum/two-sum.cpp
// Optimal Code
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        // dictionary to store: number -> index
        unordered_map<int, int> mp;
        // Traverse the array
        for (int i = 0; i < nums.size(); i++) {
            // calculate the number needed to reach the target
            int complement = target - nums[i];
            // if the complement has already been seen return its index and the current index
            if (mp.find(complement) != mp.end()) {
                return {mp[complement], i};
            }
            // store the current number along with its index
            mp[nums[i]] = i;
        }
        // no valid pair found
        return {};
    }
};

// Time Complexity : O(N)
// Space Complexity : O(1)